Doc/library/typing.rst

```diff
@@ -933,6 +933,55 @@ These can be used as types in annotations using ``[]``, each having a unique syn
 
    .. versionadded:: 3.9
 
+
+.. data:: TypeGuard
+
+   Special typing form used to annotate the return type of a user-defined
+   type guard function.  ``TypeGuard`` only accepts a single type argument.
+
+   ``TypeGuard`` aims to benefit *type narrowing* - a technique used by static
+   type checkers to determine a more precise type of an expression within a
+   program's code flow.  Usually type narrowing is done by analyzing
+   conditional code flow and applying the narrowing to a block of code.  The
+   conditional expression here is sometimes referred to as a "type guard".
+
+   Sometimes, a type guard can be a complex checking function.
+   These type guard functions require ``TypeGuard`` to narrow their input types
+   as the static type checker usually does not have enough information to
+   statically infer them.
+
+   A ``TypeGuard`` tells the static type checker that for a given function:
+      1. The return value is a boolean.
+      2. If the return value was "truthy", the type of the input to the
```

Comment:
Let's not get into truthiness here -- let's just say ``` ``True`` ```. Also, s/was/is/.

Separately, I wonder if we need to somehow distinguish more carefully between what happens in the static checker and at runtime. The checker *makes the assumption* that a `True` result implies that the expression used as the first argument has the type given in the `TypeGuard[X]` annotation. It behooves the implementation of the type guard function to honor this assumption at runtime, but neither the type checker nor the runtime has the ability to verify that it is so -- if the function "lies", that is a runtime bug for which the checker cannot take responsibility.

Reply:
Do you mean that for the current PEP, things like:

```python

def is_fake_str_list(val: List[object]) -> TypeGuard[List[str]]:
    '''Liar!'''
    return True

def func1(val: List[object]):
    if is_str_list(val):
        # Type of ``val`` is narrowed to List[str]
        print(" ".join(val))
    else:
        # Type of ``val`` remains as List[object]
        print("Not a list of strings!")

# Kaboom!
func([1,2,3])
```

Fail at runtime instead of being caught at "compile" time?

If so, I'm not sure if there's an easy way to fix that. One way would be for the static type checker to verify ``is_fake_str_list`` recursively. But this would bring about a crop of other problems, weird edge cases, and added implementation complexity for them.

Here's some comments from the PEP that I agree with (though it's talking about why there isn't strict narrowing, not runtime vs static checkers' responsibilities):

> However, there are many ways a determined or uninformed developer can subvert type safety -- most commonly by using cast or Any. If a Python developer takes the time to learn about and implement user-defined type guards within their code, it is safe to assume that they are interested in type safety and will not write their type guard functions in a way that will undermine type safety or produce nonsensical results.

Reply:
Yes, that's what I meant. And of course there's no easy fix for that. I just want the docs to not mislead the reader into thinking that that is somehow checked.